### DISCUSSION TITLE
How a traceId created in opentelemetry?

### DISCUSSION THREAD

**Discussion by lmssy** [open-telemetry/opentelemetry-java-instrumentation#3904] [Q&A] [answered]
I'm using javaagent 1.4.1 to my server which is used as an API Gateway who receive http requests from clients and transfer to dubbo requests to backend services.

I found that many http requests received by my API Gateway, the traceIds generated by my API Gateway are the same, for example as you can see below chart, it's all the same traceId 00c0a803cb16288385756511029d0001 from 2021-08-22 ~ 2021-08-23 with different http request received, it's wired, so I want to know:

1. how a traceId is generated? UUID or something else? is there any document or code that I can see?
2. Why those request received in different thread have same traceId? 

instrumentationLibrary.name is io.opentelemetry.javaagent.jetty-8.0
The API Gateway use "Jetty Continuation" framework.

![image](https://user-images.githubusercontent.com/17894328/130389808-fb47a62a-02da-4d28-aa46-0c8a734a5654.png)

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @lmssy ,

1. Trace & span ids are generated by the [`RandomIdGenerator`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/RandomIdGenerator.java) class;
2. Maybe whatever is sending all those requests always sends the same traceId in the HTTP headers? If possible, please check the headers of incoming and outgoing API gateway messages - depending on the propagator used, e.g. for W3C tracecontext check the `traceparent` header. I suspect that there might be a process that always sends HTTP requests with the same traceId, one generated at the application start for example.

**Reply by lmssy** [reply]
Thx for your reply, 

All traceIds are generated by my API Gateway server, this server is the 'root' of a trace.
I'm wonding if there is something wrong with traceId propagating on the condition of using "Jetty Continuation" which is a Asynchronous Servlets framework,  Jetty Continuation will suspend a http request and multiplexing current thread to process another request, as you can see details in this link https://wiki.eclipse.org/Jetty/Feature/Continuations.

Could you have a look or find someone to give a investigation on this? Thanks in advance！

**Reply by lmssy** [reply]
![image](https://user-images.githubusercontent.com/17894328/130412868-b10083cc-4344-4230-b525-1dbd03c042d0.png)

I have seen a comment above about how Jetty should handles async requests . please make sure jetty instrumentation do the right thing on this scene.
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/7f37514075adf8a0fbae3dcdee30f30b4029c76b/instrumentation/servlet/servlet-common/library/src/main/java/io/opentelemetry/instrumentation/servlet/ServletHttpServerTracer.java#L30

**Reply by mateuszrzeszutek** [reply]
> I'm wonding if there is something wrong with traceId propagating on the condition of using "Jetty Continuation" which is a Asynchronous Servlets framework, Jetty Continuation will suspend a http request and multiplexing current thread to process another request, as you can see details in this link https://wiki.eclipse.org/Jetty/Feature/Continuations.

That does sound like it maybe the cause behind your situation. We don't instrument Jetty continuations, and most likely all those continuation context switches and multiplexing are completely unnoticed by the OpenTelemetry agent and it assumes that everything is part of a single operation.

> Could you have a look or find someone to give a investigation on this? Thanks in advance！

Can you create an enhancement issue for Jetty Continuation instrumentation? I'm afraid that most likely the maintainers of this project won't have time and resources to work on this, but maybe we'll find a volunteer that'll instrument this lib.

**Reply by lmssy** [reply]
Ok, I will create an enhancement issue, thank you for your help.
